handlers/proxy.go

```diff
@@ -1,201 +1,103 @@
-
-		invokeService(w, r, name, forwardReq, proxyClient)
-	}
+// FunctionLookup is a openfaas-provider proxy.BaseURLResolver that allows the
+// caller to verify that a function is resolvable.
+type FunctionLookup struct {
+	docker ServiceLister
+	// dnsrr controls  if DNSRoundRobin is used to resolve function
+	dnsrr bool
+	// scheme is the http scheme (http/https) used to proxy the request
+	scheme string
+	// dnsrrLookup method used to resolve the function IP address, defaults to net.LookupIP
+	dnsrrLookup func(string) ([]net.IP, error)
 }
 
-func lookupSwarmService(serviceName string, c *client.Client) (bool, error) {
-	// fmt.Printf("Resolving: '%s'\n", serviceName)
-	serviceFilter := filters.NewArgs()
-	serviceFilter.Add("name", serviceName)
-	services, err := c.ServiceList(context.Background(), types.ServiceListOptions{Filters: serviceFilter})
-
-	return len(services) > 0, err
+// NewFunctionLookup creates a new FunctionLookup resolver
+func NewFunctionLookup(client ServiceLister, dnsrr bool) *FunctionLookup {
+	return &FunctionLookup{
+		docker:      client,
+		dnsrr:       dnsrr,
+		scheme:      urlScheme,
+		dnsrrLookup: net.LookupIP,
+	}
 }
 
-func invokeService(w http.ResponseWriter, r *http.Request, service string, forwardReq requests.ForwardRequest, proxyClient *http.Client) {
-	stamp := strconv.FormatInt(time.Now().Unix(), 10)
+// Resolve implements the openfaas-provider proxy.BaseURLResolver interface. In
+// short it verifies that a function with the given name is resolvable by Docker
+// Swarm.  It can be configured to do this via DNS or by querying the Docker Service
+// list.
+func (l *FunctionLookup) Resolve(name string) (u url.URL, err error) {
 
-	//TODO: inject setting rather than looking up each time.
-	var dnsrr bool
-	if os.Getenv("dnsrr") == "true" {
-		dnsrr = true
-	}
-
-	addr := service
-	// Use DNS-RR via tasks.servicename if enabled as override, otherwise VIP.
-	if dnsrr {
-		entries, lookupErr := net.LookupIP(fmt.Sprintf("tasks.%s", service))
-		if lookupErr == nil && len(entries) > 0 {
-			index := randomInt(0, len(entries))
-			addr = entries[index].String()
-		}
+	if l.dnsrr {
+		u.Host, err = l.byDNSRoundRobin(name)
+	} else {
+		u.Host, err = l.byName(name)
 	}
 
-	url := forwardReq.ToURL(addr, watchdogPort)
-
-	contentType := r.Header.Get("Content-Type")
-	fmt.Printf("[%s] Forwarding request [%s] to: %s\n", stamp, contentType, url)
-
-	if r.Body != nil {
-		defer r.Body.Close()
+	if err != nil {
+		return u, err
 	}
 
-	request, err := http.NewRequest(r.Method, url, r.Body)
+	u.Scheme = l.scheme
+	return u, nil
+}
 
-	copyHeaders(&request.Header, &r.Header)
+// resolve the function by checking the available docker VIP based resolution
+func (l *FunctionLookup) byName(name string) (string, error) {
+	serviceFilter := filters.NewArgs()
+	serviceFilter.Add("name", name)
+	services, err := l.docker.ServiceList(context.Background(), types.ServiceListOptions{Filters: serviceFilter})
```

Comment:
I'm not an expert using the golang context, but struck me that this will never cancel, wonder if it's better to use a context with a deadline or timeout? https://golang.org/pkg/context/#WithTimeout

Reply:
To help make this smoother, I will mimic the sql package and add a `ResolveContext` method that allows you to pass the context down.  And then move the `context.Background` to the `ResolveMethod`.  We can then decide later what the context should be and it should simplify the transition. My next commit will have this change

Reply:
If we like this, we should probably consider updating the interface in faas-provider to also have context support

Reply:
For reference on what I mean about mimicking the sql package, see https://golang.org/src/database/sql/sql.go?s=21695:21721#L738 and https://golang.org/src/database/sql/sql.go?s=21695:21721#L726

Almost all of the methods have this double implementation where the actual implementation in in the `FooContext` and then `Foo` just calls `FooContext` with a `context.Background`